### PR TITLE
Adding mvdr_beamforming FPGA reference design

### PR DESCRIPTION
**Pull request by mdbtucker**
Signed-off-by: mtucker <mike.d.b.tucker@intel.com>

# Description

Add a new DPC++ FPGA Reference Design: MVDR Beamforming.

## Type of change
- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
- [X] Command Line
- [ ] oneapi-cli
- [ ] Visual Studio
- [ ] Eclipse IDE
- [ ] VSCode

### CODE REVIEW

**Review by JoeOster — DISMISSED** [review]
I made a few changes to the readme, please take a look

Also we need to add a TPP file to the root of the sample - Please use this one https://github.com/oneapi-src/oneAPI-samples/blob/master/third-party-programs.txt

**Review by tomlenth — APPROVED** [review]

**Review by tomlenth — DISMISSED** [review]
looks good, thanks

**Comment by mdbtucker** [comment]
> I made a few changes to the readme, please take a look
> 
> Also we need to add a TPP file to the root of the sample - Please use this one https://github.com/oneapi-src/oneAPI-samples/blob/master/third-party-programs.txt

Thanks Joe.  TPP file was added (although as we've discussed I've emailed legal to protest :-) ).  Your changes to the readme look good to me.

**Review by AlbyB — DISMISSED** [review]
Looks good to me

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/mvdr_beamforming.vcxproj [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    <ClCompile Include="src\mvdr_beamforming.cpp" />
+  </ItemGroup>
+  <ItemGroup>
+    <ClInclude Include="src\BackwardSubstitution.hpp" />
+    <ClInclude Include="src\Beamformer.hpp" />
+    <ClInclude Include="src\CalcWeights.hpp" />
+    <ClInclude Include="src\Constants.hpp" />
+    <ClInclude Include="src\FakeIOPipes.hpp" />
+    <ClInclude Include="src\Substitution.hpp" />
+    <ClInclude Include="src\LoopBackKernel.hpp" />
```
Comment (tyoungsc) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/mvdr_beamforming.vcxproj:23]: This line should be deleted.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/mvdr_beamforming.vcxproj [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    <ClInclude Include="src\mvdr_complex.hpp" />
+    <ClInclude Include="src\MVDR.hpp" />
+    <ClInclude Include="src\NullPipe.hpp" />
+    <ClInclude Include="src\pipe_array_internal.hpp" />
+    <ClInclude Include="src\pipe_array.hpp" />
+    <ClInclude Include="src\PipeAggregator.hpp" />
+    <ClInclude Include="src\SteeringVectorGenerator.hpp" />
+    <ClInclude Include="src\StreamingQRD.hpp" />
+    <ClInclude Include="src\Transpose.hpp" />
+    <ClInclude Include="src\UDP.hpp" />
```
Comment (tyoungsc) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/mvdr_beamforming.vcxproj:33]: This line can also be deleted (file not added).

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/mvdr_beamforming.vcxproj (hunk 1 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/mvdr_beamforming.vcxproj]: Fixed.  Also added some missing include files.  Include file list in this file should be correct now.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/mvdr_beamforming.vcxproj (hunk 2 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/mvdr_beamforming.vcxproj]: fixed

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+## Key Implementation Details
+This reference design code sample leverages concepts that are discussed in the following FPGA tutorials: 
+* **IO Streaming** (io_streaming)
+* **Explicit Pipelining with `fpga_reg`** (fpga_register)
+* **Loop `ivdep` Attribute** (loop_ivdep)
+* **Triangular Loop Optimization** (triangular_loop)
+* **Unrolling Loops** (loop_unroll)
+* **Pipe Arrays** (pipe_array)
+
+Notably, we strongly suggest reviewing the **IO Streaming** code sample as this reference design is a direct extension of the concepts it describes.
```
Comment (akertesz) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md:29]: I'm really surprised not to see a brief discussion of "faking" IO streaming here! I don't think that a reference back to the IO streaming tutorial is sufficient -- I think it's worth a brief blurb to explain that the data in this design is actually coming from the host, that this is a handy development technique, and to see IO streaming for more details.

Update -- I found this at the bottom of the README with the design details. I suggest moving that here as it's a pretty key piece of information for anyone reading (or thinking of running) this sample.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/mvdr_beamforming.vcxproj.user [hunk 4]
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (akertesz) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/mvdr_beamforming.vcxproj.user:1]: Is this file really necessary? Almost all code samples have deleted this file, as it was unnecessary and adds clutter.

**Review by akertesz — COMMENTED** [review]
This review excludes the actual C++ code.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+* [oneAPI Programming Guide](https://software.intel.com/en-us/oneapi-programming-guide) is a general resource for target-independent DPC++ programming. 
+ 
+| Optimized for                     | Description
+---                                 |---
+| OS                                | Linux* Ubuntu* 18.04; Windows* 10
+| Hardware                          | Intel&reg; Programmable Acceleration Card (PAC) with Intel Arria&reg; 10 GX FPGA <br> Intel&reg; FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix&reg; 10 SX) <br> Intel Xeon&reg; CPU E5-1650 v2 @ 3.50GHz (host machine)
+| Software                          | Intel&reg; oneAPI DPC++ Compiler <br> Intel&reg; FPGA Add-On for oneAPI Base Toolkit
+| What you will learn               | How to perform IO streaming in oneAPI using the IO pipes SYCL extension.
+| Time to complete                  | 1 hour
+
```
[the remark this answers is not in the document]
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md:16]: Performance wasn't the main goal of this reference design, it's main purpose is to demonstrate a realistic system designed for IO streaming.  The design has not been fmax-optimized to the level of most of our reference designs, and as such I'm not sure the performance is sufficiently impressive to justify documenting here.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
+# MVDR Beamforming
+This reference design demonstrates FPGA IO streaming in oneAPI.
```
[the remark this answers is not in the document]
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md:2]: Rephrased.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+* [oneAPI DPC++ FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide) is the reference manual for targeting FPGAs through DPC++. <br>
+* [oneAPI Programming Guide](https://software.intel.com/en-us/oneapi-programming-guide) is a general resource for target-independent DPC++ programming. 
+ 
+| Optimized for                     | Description
+---                                 |---
+| OS                                | Linux* Ubuntu* 18.04; Windows* 10
+| Hardware                          | Intel&reg; Programmable Acceleration Card (PAC) with Intel Arria&reg; 10 GX FPGA <br> Intel&reg; FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix&reg; 10 SX) <br> Intel Xeon&reg; CPU E5-1650 v2 @ 3.50GHz (host machine)
+| Software                          | Intel&reg; oneAPI DPC++ Compiler <br> Intel&reg; FPGA Add-On for oneAPI Base Toolkit
+| What you will learn               | How to perform IO streaming in oneAPI using the IO pipes SYCL extension.
```
[the remark this answers is not in the document]
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md:14]: Rephrased.  Changed SYCL and oneAPI to DPC++.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+| Time to complete                  | 1 hour
+
+## Purpose
+The purpose of this tutorial is to implement a high-performance streaming IO design using oneAPI. In this reference design, we implement an MVDR-beamforming algorithm using oneAPI. More details on the algorithm can be found in the [Additional Design Information](#additional-design-information) Section.
+
+## Key Implementation Details
+This reference design code sample leverages concepts that are discussed in the following FPGA tutorials: 
+* **IO Streaming** (io_streaming)
+* **Explicit Pipelining with `fpga_reg`** (fpga_register)
```
[the remark this answers is not in the document]
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md:23]: fixed

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+| Optimized for                     | Description
+---                                 |---
+| OS                                | Linux* Ubuntu* 18.04; Windows* 10
+| Hardware                          | Intel&reg; Programmable Acceleration Card (PAC) with Intel Arria&reg; 10 GX FPGA <br> Intel&reg; FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix&reg; 10 SX) <br> Intel Xeon&reg; CPU E5-1650 v2 @ 3.50GHz (host machine)
+| Software                          | Intel&reg; oneAPI DPC++ Compiler <br> Intel&reg; FPGA Add-On for oneAPI Base Toolkit
+| What you will learn               | How to perform IO streaming in oneAPI using the IO pipes SYCL extension.
+| Time to complete                  | 1 hour
+
+## Purpose
+The purpose of this tutorial is to implement a high-performance streaming IO design using oneAPI. In this reference design, we implement an MVDR-beamforming algorithm using oneAPI. More details on the algorithm can be found in the [Additional Design Information](#additional-design-information) Section.
```
[the remark this answers is not in the document]
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md:18]: Changed tutorial to reference design.  Changed oneAPI to DPC++.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+2. Compile the design through the generated `Makefile`. The following targets are provided, and they match the recommended development flow:
+
+    * Compile for emulation (fast compile time, targets emulated FPGA device).
+
+       ```
+       make fpga_emu
+       ```
+
+    * Generate HTML performance report. Find the report in `mvdr_beamforming_report.prj/reports/report.html` directory.
```
[the remark this answers is not in the document]
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md:75]: fixed

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (akertesz) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md:16]: This makes this design somewhat different than the rest of the FPGA Reference Designs, all of which showcase competitive performance in their respective application. This is a bit of an odd duck, as it shares characteristics of Reference Designs and of Design Patterns.

Is it your intention to further optimize the performance of this design in the future?

Please consider whether it would be worthwhile including a "Performance" section here as in the other designs, and stating what you said here.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/NullPipe.hpp [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+                            // about uninitialized variables
+    success = false;
+    return( return_value );
+  }
+
+  // Blocking read
+  static T read() {
+    T return_value = T();   // call the default constructor to avoid warnings
+                            // about uninitialized variables
+    while(1) {
```
Comment (pmpeter1) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/NullPipe.hpp:40]: Why is an infinite loop correct here?

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/Transpose.hpp [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  static_assert( k_num_cols_in % k_pipe_width == 0,
+    "k_num_cols_in must be evenly divisible by k_pipe_width" );
+
+  using PipeType = NTuple< T, k_pipe_width >;
+
+  auto e = q.submit([&](handler& h) {
+
+    h.single_task< TransposeKernelName > ( [=] {
+
+      while( 1 ) {
```
Comment (pmpeter1) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/Transpose.hpp:49]: Should this be an infinite loop or just have a different termination condition?

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/mvdr_beamforming.cpp [hunk 13]
```diff
@@ -0,0 +1,627 @@
...
+    // clear the output data
+    std::fill_n(DataOutConsumer::Data(), kDataOutSize * num_matrix_copies, 0.0);
+
+    // read the input data
+    passed &= ReadInputData(in_dir, (ComplexType *)XrxTrainingProducer::Data(),
+                            (ComplexType *)XrxDataProducer::Data(), 
+                            num_matrix_copies);
+
+    // calcluate the sin(theta) values for each steering vector
+    float degree_unit = 120.0f / (kNumSteer - 1);
```
Comment (pmpeter1) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/mvdr_beamforming.cpp:163]: Can degree_unit be constexpr?

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/mvdr_beamforming.cpp [hunk 14]
```diff
@@ -0,0 +1,627 @@
...
+    XrxTrainingProducer::Destroy(q);
+    XrxDataProducer::Destroy(q);
+    SinThetaProducer::Destroy(q);
+    DataOutConsumer::Destroy(q);
+
+  } catch (exception const &e) {
+    // Catches exceptions in the host code
+    std::cerr << "Caught a SYCL host exception:\n" << e.what() << "\n";
+    // Most likely the runtime couldn't find FPGA hardware!
+    if (e.get_cl_code() == CL_DEVICE_NOT_FOUND) {
```
Comment (pmpeter1) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/mvdr_beamforming.cpp:417]: Is there a DPC++ mechanism here so we don't have to drop to OpenCL?

**Review by pmpeter1 — CHANGES_REQUESTED** [review]
Overall very nice.

Was this formatted via 'clang-format'  some of the spacing seems strange.  Also a few places seem to have extra single blank or double blank lines that distract from the main flow.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/NullPipe.hpp (hunk 11 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/NullPipe.hpp:40]: All of the 'processing' kernels have infinite loops.  This design is a streaming design, meaning it is processing a constant stream of data from some external source (normally a network port or some other external IO, in our case we have to 'fake' this with pipes that are fed from the host since we don't have a board in the dev cloud with real IO pipes).  As such, there is no clear point when the kernel should stop.  The idea is the host program launches them and then they just process any data they receive indefinitely.

**Review by akertesz — DISMISSED** [review]

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/mvdr_beamforming.cpp (hunk 13 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/mvdr_beamforming.cpp:163]: Yes, fixed.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/mvdr_beamforming.vcxproj.user (hunk 4 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/mvdr_beamforming.vcxproj.user:1]: We need this file to specify the --in argument when invoking the executable.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/mvdr_beamforming.cpp (hunk 14 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/mvdr_beamforming.cpp:417]: This is boilder plate code from all our reference designs.  I believe the only way to determine the device not found status (which is what users are most likely to run into here, so it warrants special handling) is by checking the OpenCL  code.

**Comment by mdbtucker** [comment]
Realized I made a couple of commits to a 'test branch' I was working in instead of this main branch.  All sorted out now.

I've applied clang-format to all files.  I had to add some "NO-FORMAT" comments in the source code around some attributes to keep the code readable, and lambda functions with [[kernel-args-restrict]] applied still get mildly messed up, but I can live with the result.

I believe all comments have now been addressed.

**Review by pmpeter1 — DISMISSED** [review]

**Review by JoeOster — APPROVED** [review]
